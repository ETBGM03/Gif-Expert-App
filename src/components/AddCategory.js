import React, { useState } from "react";
import PropTypes from "prop-types";

const AddCategory = ({ setCategory }) => {
  const [input, setInput] = useState("");

  const handleChange = ({ target }) => {
    setInput(target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    console.log("handleSubmit fue llamado", input);
    if (!input || !input.trim()) return alert("Please enter a valid category");
    setCategory((prev) => [input, ...prev]);
    setInput("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <small>{input}</small>
      <h2>Add New Category</h2>
      <input type="text" value={input} onChange={handleChange} />
      <button>Add</button>
    </form>
  );
};

AddCategory.propTypes = {
  setCategory: PropTypes.func.isRequired,
};
export default AddCategory;
