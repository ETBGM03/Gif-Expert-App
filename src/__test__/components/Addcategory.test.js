import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import AddCategory from "../../components/AddCategory";

describe("Pruebas en <AddCategory />", () => {
  const setCategory = jest.fn();
  let wrapper = shallow(<AddCategory setCategory={setCategory} />);

  beforeEach(() => {
    /**
     * esta funcion lo que nos permite es limpiar todos los valores o  simulaciones las cuales haya hecho en el test anterior
     */
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategory={setCategory} />);
  });

  test("debe mostrar el componente correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  /**
   * la prueba a continuaci贸n es para hacer la simulacion de que el input cambie su
   * valor y se ejecute el evento onChange del input
   *
   * nota: es importante tener en cuenta de que en caso de que nuestro componente a probar
   * tenga mas tag inputs (como no lo es este caso), si queremos testear varios inputs podemos selecconarlos
   * mediante una clase o un id el cual le hayamos proporcionado
   */
  test("debe de cambiar el state input cuando el usuario escribe en el input", () => {
    const input = wrapper.find("input");
    const value = "Hola mundo";

    input.simulate("change", { target: { value } });
    expect(wrapper.find("small").text().trim()).toBe(value);
  });

  /**
   * simularemos todo el evento onSubmit del formulario
   */

  test("NO debe hacer el fetch sino cumple con la condici贸n del handleSubmit", () => {
    const form = wrapper.find("form");
    let preventDefault = () => {};
    form.simulate("submit", { preventDefault });
    //sino tenemos un inputValue no podemos llamar a la funci贸n setCategory
    expect(setCategory).not.toHaveBeenCalled();
  });

  /**
   * simluandp el evento onSubmit del formulario llamando a la funci贸n handleSubmit y limpiamos el input
   */
  test("debe llamar el setCategory y limpiar la caja de texto", () => {
    const input = wrapper.find("input");
    const value = "Hola mundo";
    input.simulate("change", { target: { value } });

    const form = wrapper.find("form");
    let preventDefault = () => {};
    form.simulate("submit", { preventDefault });

    expect(setCategory).toHaveBeenCalled();
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
