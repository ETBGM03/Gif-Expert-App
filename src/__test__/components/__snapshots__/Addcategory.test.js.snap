// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> debe mostrar el componente correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <small />
  <h2>
    Add New Category
  </h2>
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
  <button>
    Add
  </button>
</form>
`;
